lib/dialer/public.go

```diff
@@ -89,11 +54,38 @@ func SetTrafficClass(conn net.Conn, class int) error {
 			return e1
 		}
 		return e2
-
-	case dialerConn:
-		return SetTrafficClass(conn.Conn, class)
-
 	default:
 		return fmt.Errorf("unknown connection type %T", conn)
 	}
 }
+
+func dialContextWithFallback(ctx context.Context, fallback proxy.ContextDialer, network, addr string) (net.Conn, error) {
+	dialer, ok := proxy.FromEnvironment().(proxy.ContextDialer)
+	if !ok {
+		return nil, errUnexpectedInterfaceType
+	}
+	if dialer != proxy.Direct {
+		// Capture the existing timeout by checking the deadline
+		var timeout time.Duration
+		if deadline, ok := ctx.Deadline(); ok {
+			timeout = deadline.Sub(deadline)
```

Comment:
The idea is that we want to clone the context, to do two dial calls with two repeated deadlines.

Reply:
Isn't this a bit hackish to begin with - why not just keep a `DialTimeout` method that creates its own contexts with deadlines?

Reply:
Because DialTimeout does not deal with context cancellation.

We care about both things I think, and I think inspecting if there is a deadline on the context is ok?

Reply:
I mean like in my initial commit, with the following signature:
```
func DialTimeout(ctx context.Context, network, addr string, timeout time.Duration) (net.Conn, error)
```

It feels hacky because time.Until(deadline) does not give the actual timeout, but something in between 0 and the timeout. Also there might be a higher level timeout, which was not specifically intended for dialling - in such a context it would be wrong to reset the timeout.

Reply:
Sure, yet things like built-in http client etc, do not support this method signature, and will still use a context with a deadline, at which point you'll probably only attempt to dial over one of the possible paths (via the proxy).

Reply:
Thanks for the explanation.

According to the docs that won't reset the timeout though, as the shorter deadline of the parent and passed deadline is used:

>  WithDeadline returns a copy of the parent context with the deadline adjusted to be no later than d. If the parent's deadline is already earlier than d, WithDeadline(parent, d) is semantically equivalent to parent. [...]

Maybe we should use halve the deadline on the first dial attempt then?

Reply:
I guess we could dial both connections at the same time, and return the first one to succeed, this way using a single context.

Reply:
I changed it to dial simultaneously, but I kept the previous behaviour, in that it waits for dialing via proxy and only returns the direct connection if that fails, even if direct connection is faster.